### PR TITLE
feat: interruption events

### PR DESCRIPTION
**Pull request by amar-1995**
### Details(context, link the issue, how was the bug fixed, what does the new feature do)

-
-

### CODE REVIEW

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 1]
```diff
@@ -76,6 +76,16 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
   private handleVisibilityChange = async () => {
     if (document.visibilityState === 'visible') {
       await this.replaceTrackWith(this.settings);
+      // stopped
+      AnalyticsEventFactory.interrupionStop(this.type, {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:80]: add a conditional for the fn and then pass the params. args are same

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 2]
```diff
@@ -60,7 +60,7 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
     }
     this.pluginsManager = new HMSAudioPluginsManager(this, eventBus, room);
     this.setFirstTrackId(track.id);
-    if (isIOS() && isBrowser) {
+    if (isMobile() && isBrowser) {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:63]: this works on android. why are we changing without an understanding? this will increase testing scope

File: packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts [hunk 3]
```diff
@@ -256,6 +257,28 @@ export default class AnalyticsEventFactory {
...
+      name: 'interruption.start',
+      level: AnalyticsEventLevel.INFO,
+      properties: {
+        type,
+        ...deviceInfo,
+      },
+    });
+  }
+
+  static interrupionStop(type: HMSTrackType, deviceInfo: Partial<MediaDeviceInfo>) {
```
Comment (raviteja83) [inline packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts:272]: one event is enough. you can take a bool for start or stop and add a condition in name. both are exactly same

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts (hunk 2 above)
Comment (amar-1995) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:63]: I tested it on android, it is not working. Condition above is checking for ios browser only.

**Comment by raviteja83** [comment]
AudioSinkManager also has interuption that also has to be handled.

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts (hunk 2 above)
Comment (raviteja83) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:63]: does the change make a difference?

File: packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts (hunk 3 above)
Comment (raviteja83) [reply packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts:272]: only mark as resolved after you push the changes.

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts (hunk 2 above)
Comment (amar-1995) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:63]: I added this check which I not exactly remember. But it is only for ios device where we want to reacquire the audio.
https://github.com/100mslive/web-sdks/pull/1113/files

File: packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ import { HMSException } from '../error/HMSException';
 import { DeviceMap, SelectedDevices } from '../interfaces';
 import { HMSTrackSettings } from '../media/settings/HMSTrackSettings';
 import { HMSRemoteVideoTrack } from '../media/tracks/HMSRemoteVideoTrack';
+import { HMSTrackType } from '../schema';
```
Comment (raviteja83) [inline packages/hms-video-store/src/analytics/AnalyticsEventFactory.ts:10]: i don't think should be from schema. see if the same is there elsewhere.

File: packages/hms-video-store/src/media/tracks/HMSTrack.ts [hunk 5]
```diff
@@ -84,7 +85,16 @@ export abstract class HMSTrack {
   protected setFirstTrackId(trackId: string) {
     this.firstTrackId = trackId;
   }
-
+  /**
+   * @internal
+   * It will send event to analytics when interruption start/stop
+   */
+  sendInterruptionEvent(started: boolean, isRemoteAudio = false) {
+    AnalyticsEventFactory.interrupion(started, isRemoteAudio ? 'remote.audio' : this.type, {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSTrack.ts:93]: this will only give the event. where are we sending it?

File: packages/hms-video-store/src/audio-sink-manager/AudioSinkManager.ts [hunk 6]
```diff
@@ -116,13 +116,22 @@ export class AudioSinkManager {
...
+      audioTrack.sendInterruptionEvent(true, true);
     }
   };
 
+  private handleAudioPlay = async (event: any) => {
+    // this means the audio replayed because of external factors(headset removal, incoming phone call)
+    const audioTrack = this.store.getTrackById(event.target.id);
+    if (audioTrack) {
+      // stopped interruption event
+      audioTrack.sendInterruptionEvent(false, true);
```
Comment (raviteja83) [inline packages/hms-video-store/src/audio-sink-manager/AudioSinkManager.ts:132]: we should probably send only once, not for each track. will be too many events in a large room.

File: packages/hms-video-store/src/audio-sink-manager/AudioSinkManager.ts [hunk 7]
```diff
@@ -140,6 +149,7 @@ export class AudioSinkManager {
     audioEl.style.display = 'none';
     audioEl.id = track.trackId;
     audioEl.addEventListener('pause', this.handleAudioPaused);
+    audioEl.addEventListener('play', this.handleAudioPlay);
```
Comment (raviteja83) [inline packages/hms-video-store/src/audio-sink-manager/AudioSinkManager.ts:152]: don't this this helps much. we can mark as false when the track is played again. there is a method to replay.

File: packages/hms-video-store/src/media/tracks/HMSTrack.ts [hunk 8]
```diff
@@ -84,7 +85,16 @@ export abstract class HMSTrack {
   protected setFirstTrackId(trackId: string) {
     this.firstTrackId = trackId;
   }
-
+  /**
+   * @internal
+   * It will send event to analytics when interruption start/stop
+   */
+  sendInterruptionEvent({ started, isRemoteAudio = false }: { started: boolean; isRemoteAudio?: boolean }) {
+    return AnalyticsEventFactory.interrupion(started, isRemoteAudio ? 'remote.audio' : this.type, {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSTrack.ts:93]: this is not sent to the server

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 9]
```diff
@@ -74,8 +77,22 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
   }
 
   private handleVisibilityChange = async () => {
-    if (document.visibilityState === 'visible') {
+    if (this.nativeTrack.readyState === 'ended' && this.nativeTrack.muted && document.visibilityState !== 'visible') {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:80]: this is not necessarily true. it should be ended or muted

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 10]
```diff
@@ -74,8 +77,22 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
...
+      this.eventBus.analytics.publish(
+        this.sendInterruptionEvent({
+          started: true,
+        }),
+      );
+      this.isInterrupted = true;
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:86]: why this extra flag, add unnecessary complexity, the track status can be checked when visibile

File: packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts [hunk 11]
```diff
@@ -499,9 +499,21 @@ export class HMSLocalVideoTrack extends HMSVideoTrack {
       this.enabledStateBeforeBackground = this.enabled;
       this.nativeTrack.enabled = false;
       this.replaceSenderTrack(this.nativeTrack);
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts:501]: can we check if this is needed. not relveant to this PR but good to check if it is needed

File: packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts [hunk 12]
```diff
@@ -499,9 +499,21 @@ export class HMSLocalVideoTrack extends HMSVideoTrack {
       this.enabledStateBeforeBackground = this.enabled;
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts:499]: add an early return when source !== 'regular' otherwise for screenshare also it might go into the else.

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 13]
```diff
@@ -72,10 +69,22 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
   resetManuallySelectedDeviceId() {
     this.manuallySelectedDeviceId = undefined;
   }
-
+  // eslint-disable-next-line complexity
   private handleVisibilityChange = async () => {
-    if (document.visibilityState === 'visible') {
+    if ((this.nativeTrack.readyState === 'ended' || this.nativeTrack.muted) && document.visibilityState !== 'visible') {
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:74]: can we add the same check here for regular that is added to video.

File: packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts [hunk 14]
```diff
@@ -510,7 +510,7 @@ export class HMSLocalVideoTrack extends HMSVideoTrack {
       );
     } else {
       this.nativeTrack.enabled = this.enabledStateBeforeBackground;
-      this.replaceSenderTrack(this.processedTrack || this.nativeTrack);
+      // this.replaceSenderTrack(this.processedTrack || this.nativeTrack);
```
Comment (raviteja83) [inline packages/hms-video-store/src/media/tracks/HMSLocalVideoTrack.ts:513]: why is this removed?

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 15]
```diff
@@ -73,9 +72,30 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
...
   private handleVisibilityChange = async () => {
-    if (document.visibilityState === 'visible') {
+    // track state is fine do nothing
+    if (!this.isTrackNotPublishing()) {
+      HMSLogger.d(this.TAG, `visibiltiy: ${document.visibilityState}`, `${this}`);
+      return;
+    }
+    if (document.visibilityState === 'hidden') {
+      this.eventBus.analytics.publish(
+        this.sendInterruptionEvent({
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: We were not getting interruption events for audio track yesterday, is that working now?

Comment (raviteja83) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: you were checking in audio sink manager I think. it might work not so sure. have not tested myself.

Comment (KaustubhKumar05) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: Was checking for events on events inspector as well

Comment (amar-1995) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: @raviteja83 Audio is not working because the track state is not ended or muted.

Comment (amar-1995) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: I tried going into the background. I tried calling as well with @KaustubhKumar05 on Monday. It was not working

Comment (raviteja83) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:87]: might work on iOS.

**Review by raviteja83 — APPROVED** [review]

**Review by eswarclynn — APPROVED** [review]
